src/MenuItem/index.js

```diff
@@ -18,62 +18,82 @@ const subChevron = css.resolve`
     }
 `
 
-function SubMenu({ size, list, onClick, className }) {
-    return (
-        <div className={cx('sub-menu', className)}>
-            <Menu size={size} list={list} onClick={onClick} />
-            <style jsx>{styles}</style>
-        </div>
-    )
+const subMenu = css.resolve`
+    div {
+        display: none;
+        position: absolute;
+        top: 0;
+        left: 100%;
+        white-space: nowrap;
+    }
+
+    .item:hover > div {
+        display: block;
+    }
+`
+
+function SubMenu({ size, children, onClick, className }) {
+    return <div className={className}>{children}</div>
 }
 
 export default function MenuItem({
     label,
-    value,
     icon,
-    list,
+    children,
     active,
     disabled,
     size,
     onClick,
     className,
 }) {
-    const hasMenu = list.length > 0
+    const hasMenu = !!children
     return (
         <li
-            className={cx('item', className, { disabled, active })}
+            className={cx('item', className, subMenu.className, {
+                disabled,
+                active,
+            })}
             onClick={evt => {
                 if (onClick) {
                     evt.preventDefault()
                     evt.stopPropagation()
-                    onClick(value)
+                    onClick()
```

Comment:
On the face of it, sure, it does not look like a simplification. What I was thinking about was that the handler function for onClick becomes simpler.

The way it worked before, we would have one massive `onClick` handler which would look at the `value` and decide what to do based on it. You would typically have one handler for the entire menu.

It can be through a map or a `switch` or `if` statements, but I find it messy and annoying to work with: https://github.com/dhis2/ui-core/blob/1.0.0-beta.5/src/widgets/HeaderBar/Profile.js#L125-L141

A consequence of the old implementation was also that the `MenuItem`'s `onClick` function was added by the `Menu` component, which locked the implementation to passing in an array of objects which the `Menu` component then converted into `MenuItem`s and attached `onClick` handlers to each `MenuItem`.

So the old implementation would be:

```jsx
function handler(name, value) {
    switch (name):
        case 'foobar': {
             doSomething(value)
        }
        default: {
             console.error('RAAAH')
        }
}

const list = [
{
        label: `Menu item 1`,
        name: 'foobar',
        value: 1,
        icon: 'favorite',
},
{
        label: `Menu item 2`,
        name: 'fuzbuz',
        value: 2,
        icon: 'favorite',
}
]

<DropdownButton onClick={handler} list={options} />
```

With nested sub-menus this approach is very unwieldy, as you will have a single `onClick` handler to handle **all** menu options. So if you have sub-menus with options you will have to handle those as well on the top-level.

With the new implementation the consumer can generate the full menu and attach small `onClick` handlers **which do exactly one thing**, e.g.:

```jsx
    <Menu>
        <Divider />
        <MenuItem
            label="Two"
            value="two"
            onClick={() => {
                alert('this is no.2')
            }}
        />
        <MenuItem
            label="Three"
            value="three"
            onClick={() => {
                alert('this is no.3')
            }}
        />
    </Menu>

<DropdownButton label="Drop it with menu!" component={menu} />
```

To use nested Menu components with the new setup:

```jsx
    <Menu>
        <MenuItem label="One" value="one">
            <Menu>
                <MenuItem label="Sub One" value="subone">
                    <Menu>
                        <MenuItem label="Sub One-One" value="suboneone">
                            <Menu>
                                <MenuItem
                                    label="Sub One-One-One"
                                    value="suboneoneone"
                                    onClick={() => {
                                        alert('you went deep!')
                                    }}
                                />
                            </Menu>
                        </MenuItem>
                    </Menu>
                </MenuItem>
            </Menu>
        </MenuItem>
    </Menu>
```

With the old:

```jsx
const submenu = [
{
        label: `Sub Menu item 1`,
        name: 'foo',
        value: 4,
        icon: 'favorite',
},
{
        label: `Sub item 2`,
        name: 'fuz',
        value: 3,
        icon: 'favorite',
}
]

const list = [
{
        label: `Menu item 1`,
        name: 'foobar',
        value: 1,
        icon: 'favorite',
        list: submenu,
},
{
        label: `Menu item 2`,
        name: 'fuzbuz',
        value: 2,
        icon: 'favorite',
}
]
```

Reply:
Sorry for the confusion! I still want the `MenuItem` to be decoupled from the `Menu`!

>     <Menu>
>        <MenuItem label="One" value="one">
>            <Menu>
>                <MenuItem label="Sub One" value="subone">
>                    <Menu>
>                        <MenuItem label="Sub One-One" value="suboneone">
>                            <Menu>
>                                <MenuItem
>                                    label="Sub One-One-One"
>                                    value="suboneoneone"
>                                    onClick={() => {
>                                        alert('you went deep!')
>                                    }}
>                                />
>                            </Menu>
>                        </MenuItem>
>                    </Menu>
>                </MenuItem>
>            </Menu>
>        </MenuItem>
>    </Menu>

You removed the `value` prop, so just to clarify, the `MenuItem`s without a callback wouldn't have a value prop either in your example, right?
 
the only difference to your code is, that mine would look like this:
```jsx
    <Menu>
        <MenuItem label="One">
            <Menu>
                <MenuItem label="Sub One">
                    <Menu>
                        <MenuItem label="Sub One-One">
                            <Menu>
                                <MenuItem
                                    label="Sub One-One-One"
                                    value="suboneoneone"
                                    onClick={(value) => {
                                        alert(`you went deep with ${value}!`)
                                    }}
                                />
                            </Menu>
                        </MenuItem>
                    </Menu>
                </MenuItem>
            </Menu>
        </MenuItem>
    </Menu>
```

Having said that, the code examples you provide make it clear that using components that way is very complicated. But it's - in my opinion - an antipattern to what our application architecture should look like.

Clicking on an item should finally trigger an action, resulting in a state change (with a possible derived state, handled by a reselect selector). Everything else would be mixing business logic with view logic!

If the callback is for handling UI state and is passed down from a parent component and is as complicated as your example, 1. there's a code smell and 2. having a method (passed down as callback) in the parent component for every scenario seems like an anti pattern to me.

Reply:
Ah! I see what you mean. Yes, there should be a `value` prop and parameter!

Reply:
f8bcab8